File: static/python/smallest_subarray_covering_set.py
import sys
import collections
import random
from smallest_subarray_covering_set_stream import find_smallest_subarray_covering_subset
import string


# @include
def find_smallest_subarray_covering_set(paragraph, keywords):
    keywords_to_cover = collections.Counter(keywords)
    result = (-1, -1)
    remaining_to_cover = len(keywords)
    left = 0
    for right, p in enumerate(paragraph):
        if p in keywords:
            keywords_to_cover[p] -= 1
            if keywords_to_cover[p] >= 0:
                remaining_to_cover -= 1

        # Keeps advancing left until keywords_to_cover does not contain all
        # keywords.
        while remaining_to_cover == 0:
            if result == (-1, -1) or right - left < result[1] - result[0]:
                result = (left, right)
            pl = paragraph[left]
            if pl in keywords:
                keywords_to_cover[pl] += 1
                if keywords_to_cover[pl] > 0:
                    remaining_to_cover += 1
            left += 1
    return result
# @exclude


def rand_string(length):
    return ''.join(random.choice(string.ascii_lowercase) for _ in range(length))


# O(n^2) solution
def check_ans(A, Q):
    d = set(Q)
    ans = (0, len(A) - 1)
    for l in range(len(A)):
        count = set()
        for r in range(l, len(A)):
            if r - l >= ans[1] - ans[0]:
                break
            if A[r] in d:
                count.add(A[r])
            if len(count) == len(Q):
                if r - l < ans[1] - ans[0]:
                    ans = (l, r)
                break
    return ans[1] - ans[0]


def simple_test_case(A, d, start, finish):
    res = find_smallest_subarray_covering_set(A, set(d))
    print('res =', *res)
    assert res == (start, finish)
    res = find_smallest_subarray_covering_subset(iter(A), d)
    print('res =', *res)
    assert res == (start, finish)


def simple_test():
    A = ['a', 'b', 'c', 'b', 'a', 'd', 'c', 'a', 'e', 'a', 'a', 'b', 'e']
    d = ['b', 'c', 'e']
    simple_test_case(A, d, 3, 8)
    d = ['a', 'c']
    simple_test_case(A, d, 6, 7)
    A = ['a', 'b']
    d = ['a', 'b']
    simple_test_case(A, d, 0, 1)
    A = ['a', 'b']
    d = ['b', 'a']
    simple_test_case(A, d, 0, 1)


def main():
    simple_test()
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)
        A = [rand_string(random.randint(1, 10)) for _ in range(n)]
        d = set(A)
        m = random.randint(1, len(d))
        Q = random.sample(d, m)
        inp = set(Q)
        res = find_smallest_subarray_covering_set(A, inp)
        print(*res, sep=', ')
        assert not inp.difference(A[res[0]:res[1] + 1])
        res2 = find_smallest_subarray_covering_subset(iter(A), Q)
        print(*res2, sep=', ')
        assert not inp.difference(A[res[0]:res[1] + 1])
        assert res[1] - res[0] == res2[1] - res2[0]
        assert res[1] - res[0] == check_ans(A, Q)


if __name__ == '__main__':
    main()

File: static/cpp/java/src/main/java/com/epi/SmallestSubarrayCoveringSet.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.io.IOException;

public class SmallestSubarrayCoveringSet {
  public static String randString(int len) {
    StringBuilder sb = new StringBuilder();
    Random gen = new Random();
    while (len-- != 0) {
      sb.append((char)(gen.nextInt('z' + 1 - 'a') + 'a'));
    }
    return sb.toString();
  }

  // @include
  // Represent subarray by starting and ending indices, inclusive.
  private static class Subarray {
    public Integer start;
    public Integer end;

    public Subarray(Integer start, Integer end) {
      this.start = start;
      this.end = end;
    }
  }

  public static Subarray findSmallestSubarrayCoveringSet(List<String> paragraph,
                                                         Set<String> keywords) {
    Map<String, Integer> keywordsToCover = new HashMap<>();
    for (String keyword : keywords) {
      keywordsToCover.put(keyword, keywordsToCover.containsKey(keyword)
                                       ? keywordsToCover.get(keyword) + 1
                                       : 1);
    }

    Subarray result = new Subarray(-1, -1);
    int remainingToCover = keywords.size();
    for (int left = 0, right = 0; right < paragraph.size(); ++right) {
      Integer keywordCount = keywordsToCover.get(paragraph.get(right));
      if (keywordCount != null) {
        keywordsToCover.put(paragraph.get(right), --keywordCount);
        if (keywordCount >= 0) {
          --remainingToCover;
        }
      }

      // Keeps advancing left until it reaches end or keywordsToCover does not
      // have all keywords.
      while (remainingToCover == 0) {
        if ((result.start == -1 && result.end == -1)
            || right - left < result.end - result.start) {
          result.start = left;
          result.end = right;
        }
        keywordCount = keywordsToCover.get(paragraph.get(left));
        if (keywordCount != null) {
          keywordsToCover.put(paragraph.get(left), ++keywordCount);
          if (keywordCount > 0) {
            ++remainingToCover;
          }
        }
        ++left;
      }
    }
    return result;
  }
  // @exclude

  // O(n^2) solution
  public static int checkAns(List<String> A, List<String> Q) {
    Set<String> dict = new HashSet<>(Q);

    Subarray ans = new Subarray(0, A.size() - 1);
    for (int l = 0; l < A.size(); ++l) {
      Map<String, Integer> count = new HashMap<>();
      for (int r = l; r < A.size() && r - l < ans.end - ans.start; ++r) {
        if (dict.contains(A.get(r))) {
          count.put(A.get(r),
                    count.containsKey(A.get(r)) ? count.get(A.get(r)) + 1 : 1);
        }
        if (count.size() == Q.size()) {
          if (r - l < ans.end - ans.start) {
            ans.start = l;
            ans.end = r;
          }
          break;
        }
      }
      count.clear();
    }

    return ans.end - ans.start;
  }

  private static void simpleTestCase(List<String> A, List<String> dict,
                                     int start, int finish)
      throws ClassNotFoundException, IOException {
    Subarray res
        = findSmallestSubarrayCoveringSet(A, new HashSet<String>(dict));
    System.out.println("res = " + res.start + " " + res.end);
    assert(res.start == start && res.end == finish);
    Integer[] res2Array
        = SmallestSubarrayCoveringSetStream
              .findSmallestSubarrayCoveringSubsetAsArray(A, dict);
    System.out.println("res2Array = " + res2Array[0] + " " + res2Array[1]);
    assert(res2Array[0] == start && res2Array[1] == finish);
  }

  private static void simpleTest() throws ClassNotFoundException, IOException {
    List<String> A = Arrays.asList("a", "b", "c", "b", "a", "d", "c", "a", "e",
                                   "a", "a", "b", "e");
    List<String> dict = Arrays.asList("b", "c", "e");
    simpleTestCase(A, dict, 3, 8);
    dict = Arrays.asList("a", "c");
    simpleTestCase(A, dict, 6, 7);
    A = Arrays.asList("a", "b");
    dict = Arrays.asList("a", "b");
    simpleTestCase(A, dict, 0, 1);
    A = Arrays.asList("a", "b");
    dict = Arrays.asList("b", "a");
    simpleTestCase(A, dict, 0, 1);
  }

  public static void main(String[] args)
      throws ClassNotFoundException, IOException {
    simpleTest();
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = gen.nextInt(10000) + 1;
      }
      List<String> A = new ArrayList<>(n);
      for (int i = 0; i < n; ++i) {
        A.add(randString(gen.nextInt(10) + 1));
      }
      Set<String> dict = new HashSet<>(A);
      int m = gen.nextInt(dict.size()) + 1;
      List<String> Q = new ArrayList<>(m);
      int idx = 0;
      for (String aDict : dict) {
        Q.add(aDict);
        if (--m == 0) {
          break;
        }
      }

      dict = new HashSet<>(Q);
      Subarray res = findSmallestSubarrayCoveringSet(A, dict);
      System.out.println(res.start + ", " + res.end);
      dict.clear();
      for (String aQ1 : Q) {
        dict.add(aQ1);
      }
      for (int i = res.start; i <= res.end; ++i) {
        if (dict.contains(A.get(i))) {
          dict.remove(A.get(i));
        }
      }
      assert(dict.isEmpty());
      Integer[] res2Array
          = SmallestSubarrayCoveringSetStream
                .findSmallestSubarrayCoveringSubsetAsArray(A, Q);
      Subarray res2 = new Subarray(res2Array[0], res2Array[1]);

      System.out.println(res2.start + ", " + res2.end);
      dict.clear();
      for (String aQ : Q) {
        dict.add(aQ);
      }
      for (int i = res.start; i <= res.end; ++i) {
        if (dict.contains(A.get(i))) {
          dict.remove(A.get(i));
        }
      }
      assert(dict.isEmpty());
      assert(res.end - res.start == res2.end - res2.start);
      assert(res.end - res.start == checkAns(A, Q));
    }
  }
}
